### PR TITLE
Create a client that can work with the cardano node.

### PR DESCRIPTION
**Pull request by raduom**
I moved the current mocked server and client to a `Mock` package, and I removed the Server from the main package.

Switched to monomorphic types, but will think if it makes sense to parameterise more on the next refactoring rounds.

To test it, you need to start the testnet client and then connect to it using ```cabal exec sync-client -- <path-to-socket>```.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-pab/src/Cardano/Node/Types.hs [hunk 1]
```diff
@@ -75,6 +76,37 @@ import qualified Wallet.Emulator.MultiAgent     as MultiAgent
 
 import           Ledger.Fee                     (FeeConfig)
 import           Plutus.PAB.Arbitrary           ()
+-- =======
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Node/Types.hs:79]: merge artifacts?

File: plutus-pab/src/Cardano/Protocol/Socket/Client.hs [hunk 2]
```diff
@@ -69,7 +68,11 @@ runChainSync socketPath slotConfig onNewBlock = do
     _ <- forkIO $ withIOManager $ loop (1 :: Second) handle
     pure handle
     where
-      loop :: TimeUnit a => a -> ChainSyncHandle -> IOManager -> IO ()
+      loop :: forall a. TimeUnit a
+           => a
+           -> (ChainSyncHandle CBlock)
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Client.hs:73]:
```suggestion
           -> ChainSyncHandle CBlock
```

File: plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+import           Ouroboros.Network.Mux
+import           Ouroboros.Network.NodeToClient                      (NodeToClientProtocols (..), connectTo,
+                                                                      versionedNodeToClientProtocols)
+import           Ouroboros.Network.Snocket
+import           Ouroboros.Network.Socket
+
+import           Cardano.Protocol.Socket.Type
+import           Ledger                                              (Block, Slot (..), Tx (..))
+
+data ChainSyncHandle = ChainSyncHandle
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs:34]: Can we use the `ChainSyncHandle` from `Cardano.Protocol.Socket.Client` instead?

File: plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+ -> Tx
+ -> IO ()
+queueTx TxSendHandle { tshQueue } tx =
+    atomically (writeTQueue tshQueue tx)
+
+getCurrentSlot :: ChainSyncHandle -> IO Slot
+getCurrentSlot = cshCurrentSlot
+
+-- | Run the chain sync protocol to get access to the current slot number.
+runChainSync' :: FilePath
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs:54]: should probably be called `runChainSyncMock'` (same for `runChainSync` below), just to make it abundantly clear that these are the emulator txn types

File: plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs [hunk 5]
```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE TypeFamilies      #-}
+
+module Cardano.Protocol.Socket.Mock.Client where
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs:8]: Needs some comments, at least a pointer to the non-mock client, and explain what the difference is

File: plutus-pab/src/Cardano/Protocol/Socket/Type.hs [hunk 6]
```diff
@@ -11,22 +12,33 @@ module Cardano.Protocol.Socket.Type where
...
 import           GHC.Generics
 import           NoThunks.Class                                     (NoThunks)
 
+import           Cardano.Chain.Slotting                             (EpochSlots (..))
 import           Codec.Serialise                                    (DeserialiseFailure)
 import qualified Codec.Serialise                                    as CBOR
 import           Network.TypedProtocol.Codec
+import qualified Ouroboros.Consensus.Byron.Ledger                   as Byron
+import           Ouroboros.Consensus.Cardano.Block                  (CardanoBlock, CodecConfig (..))
+-- import           Ouroboros.Consensus.Network.NodeToClient           (ClientCodecs, Codecs, Codecs' (..), clientCodecs)
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Type.hs:36]: commented out

File: plutus-pab/src/Cardano/Protocol/Socket/Type.hs [hunk 7]
```diff
@@ -89,7 +101,7 @@ nodeToClientVersion = NodeToClientV_4
 -- argument to the client connection function in a future PR (the network magic
 -- number matches the one in the test net created by scripts)
 nodeToClientVersionData :: NodeToClientVersionData
-nodeToClientVersionData = NodeToClientVersionData { networkMagic = NetworkMagic 42 }
+nodeToClientVersionData = NodeToClientVersionData { networkMagic = NetworkMagic 1097911063 }
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Type.hs:104]: is this for some testnet?

File: plutus-pab/src/Cardano/Protocol/Socket/Type.hs [hunk 8]
```diff
@@ -107,19 +119,51 @@ doNothingResponderProtocol =
 type Offset = Integer
 
 -- | Boilerplate codecs used for protocol serialisation.
-codecChainSync :: Codec (ChainSync.ChainSync Block (Point Block) Tip)
-                        DeserialiseFailure
-                        IO BSL.ByteString
-codecChainSync =
+epochSlots :: EpochSlots
+epochSlots = EpochSlots 432000
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Protocol/Socket/Type.hs:123]: where does this number come from?

File: plutus-pab/src/Cardano/Wallet/Mock.hs [hunk 9]
```diff
@@ -69,6 +70,44 @@ import           Wallet.Emulator.LogMessages      (TxBalanceMsg)
 import           Wallet.Emulator.NodeClient       (emptyNodeClientState)
 import           Wallet.Emulator.Wallet           (Wallet (..), WalletState (..), defaultSigningProcess)
 import qualified Wallet.Emulator.Wallet           as Wallet
+-- =======
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Wallet/Mock.hs:73]: merge artifacts

File: plutus-pab/src/Cardano/Wallet/Mock.hs [hunk 10]
```diff
@@ -18,20 +18,21 @@ module Cardano.Wallet.Mock
...
+                                                      WalletMsg (..), Wallets)
+import           Control.Concurrent                  (MVar)
+import           Control.Concurrent.MVar             (putMVar, takeMVar)
+import           Control.Lens                        (at, (.~))
+import           Control.Monad.Error                 (MonadError)
+import qualified Control.Monad.Except                as MonadError
 import           Control.Monad.Freer
 import           Control.Monad.Freer.Error
 import           Control.Monad.Freer.Extras
+-- <<<<<<< HEAD
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Wallet/Mock.hs:35]: merge artifacts

File: plutus-pab/src/Cardano/Wallet/Server.hs [hunk 11]
```diff
@@ -12,12 +12,13 @@ module Cardano.Wallet.Server
...
-import           Cardano.ChainIndex.Types         (ChainIndexUrl (..))
-import qualified Cardano.Protocol.Socket.Client   as Client
-import           Cardano.Wallet.API               (API)
+import           Cardano.BM.Data.Trace               (Trace)
+import qualified Cardano.ChainIndex.Client           as ChainIndexClient
+import           Cardano.ChainIndex.Types            (ChainIndexUrl (..))
+import qualified Cardano.Protocol.Socket.Mock.Client as Client
+import           Cardano.Wallet.API                  (API)
 import           Cardano.Wallet.Mock
+-- <<<<<<< HEAD
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Wallet/Server.hs:21]: merge artifacts

File: plutus-pab/src/Cardano/Wallet/Server.hs [hunk 12]
```diff
@@ -44,6 +45,33 @@ import           Wallet.Effects                   (balanceTx, ownPubKey, startWa
                                                    walletAddSignature)
 import           Wallet.Emulator.Wallet           (Wallet (..), emptyWalletState)
 import qualified Wallet.Emulator.Wallet           as Wallet
+-- =======
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Wallet/Server.hs:48]: merge artifacts

File: plutus-pab/sync-client/Main.hs [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+-- | We only need to know the location of the socket.
+--   We can get the protocol versions from Cardano.Protocol.Socket.Type
+data Configuration = Configuration
+  { cSocketPath :: String
+  } deriving (Show)
+
+slotConfig :: SlotConfig
+slotConfig =
+  SlotConfig
+    { scZeroSlotTime = 1591566291000
```
Comment (j-mueller) [inline plutus-pab/sync-client/Main.hs:27]: where does this number come from?

File: plutus-pab/sync-client/Main.hs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+main = do
+  cfg <- execParser $ info (cfgParser <**> helper)
+                        ( fullDesc
+                        <> progDesc "Connect and stream blocks from the node client"
+                        )
+  putStrLn "Runtime configuration:"
+  putStrLn "----------------------"
+  print    cfg
+  _ <- runChainSync (cSocketPath cfg) slotConfig processBlock
+  _ <- forever $ threadDelay 43
```
Comment (j-mueller) [inline plutus-pab/sync-client/Main.hs:84]: 43?

**Review by j-mueller — COMMENTED** [review]
Generally looking good. We can use the same client in the chain index, which is nice. Just needs some tidying / comments.

File: plutus-pab/sync-client/Main.hs (hunk 14 above)
Comment (raduom) [reply plutus-pab/sync-client/Main.hs:84]: I missed that. Should have been 42. Obi-one bug.

File: plutus-pab/src/Cardano/Node/Types.hs (hunk 1 above)
Comment (raduom) [reply plutus-pab/src/Cardano/Node/Types.hs:79]: Yeah. I reviewed the code before finding out there were conflicts.

File: plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs (hunk 4 above)
Comment (raduom) [reply plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs:54]: I thought it was cool using the same name from different packages, but can do.

Comment (j-mueller) [reply plutus-pab/src/Cardano/Protocol/Socket/Mock/Client.hs:54]: True. The risk of someone confusing the two probably isn't that high. So I'm happy to leave it as it is.

File: plutus-pab/src/Cardano/Protocol/Socket/Type.hs (hunk 7 above)
Comment (raduom) [reply plutus-pab/src/Cardano/Protocol/Socket/Type.hs:104]: Yes.

Comment (raduom) [reply plutus-pab/src/Cardano/Protocol/Socket/Type.hs:104]: For *the* testnet actually.

File: plutus-pab/src/Cardano/Protocol/Socket/Type.hs (hunk 8 above)
Comment (raduom) [reply plutus-pab/src/Cardano/Protocol/Socket/Type.hs:126]: Added a comment.

File: plutus-pab/sync-client/Main.hs (hunk 13 above)
Comment (raduom) [reply plutus-pab/sync-client/Main.hs:29]: Added a comment.

**Review by j-mueller — APPROVED** [review]
:shipit:

**Pull request merged**